### PR TITLE
Approve LH device for user

### PR DESCRIPTION
**Pull request by ChrisPenner**
Add LH client to a user when they approve it.

### `/teams/{tid}/legalhold/{uid}/approve`
- [x] Permissions: only the user themself can call this.
- [x] fail if no legal hold service registered
- [x] fail if legal Hold feature flag disabled
- [ ] fail if GLOBAL legal Hold feature flag disabled
- [x] PUT:
    - [x] generate and store legalhold (user) tokens/cookies to pass on to the service (⚠️ currently full-auth tokens from sso-login)
    - [x] add LH device to user's list of devices
    - [ ] ⚠️ Do we want this?: send an event to all user's clients (and possibly also to all admin's clients? (to show in the admin panel?))

### CODE REVIEW

File: libs/bilge/src/Bilge/Response.hs [hunk 1]
```diff
@@ -42,6 +45,19 @@ getHeader' h = fromMaybe "NO_HEADER_VALUE" . getHeader h
 getCookie :: ByteString -> Response a -> Maybe Cookie
 getCookie n r = find ((n ==) . cookie_name) (destroyCookieJar $ responseCookieJar r)
 
+-- | Retrieve the value of a given cookie name from a "Set-Cookie" header on the response
+getCookieValue :: ByteString -> Response a -> Maybe ByteString
```
Comment (ChrisPenner) [inline libs/bilge/src/Bilge/Response.hs]: `getCookie` didn't work for me at all; From the docs it seems like you need to convince it to use the cookie jar somehow? I tried setting a Cookie Jar on the request but it didn't do the trick. Maybe someone can help me with this?

File: services/galley/src/Galley/Intra/Client.hs [hunk 2]
```diff
@@ -40,3 +51,51 @@ notifyClientsAboutLegalHoldRequest requesterUid targetUid lastPrekey' prekeys =
...
+           method POST
+            . host brigHost
+            . port brigPort
+            . path "/i/sso-login" -- ^ TODO: switch to '/i/legalhold-login'
+            . queryItem "persist" "true"
+            . json (SsoLogin uid Nothing)
+            . collectCookies
+            . expect2xx
+    lg <- view applog
+    case getCookieValue "zuid" r of
```
Comment (ChrisPenner) [inline services/galley/src/Galley/Intra/Client.hs]: I tried setting a custom cookie label but it didn't work; maybe that's something other than what I thought? At any rate, using `zuid` works.

This will all change when we switch to the new token endpoint so don't get too hung up on it 😋

File: services/galley/test/integration/API/Util.hs [hunk 3]
```diff
@@ -748,20 +748,6 @@ isMember usr cnv = do
                . expect2xx
     return $ isJust (decodeBody res :: Maybe Member)
 
-someLastPrekeys :: [LastPrekey]
```
Comment (ChrisPenner) [inline services/galley/test/integration/API/Util.hs]: Just moved to the bottom 👍

File: services/galley/test/integration/TestHelpers.hs [hunk 4]
```diff
@@ -13,10 +13,7 @@ import qualified API.SQS             as SQS
 test :: IO TestSetup -> TestName -> TestM a -> TestTree
 test s n h = testCase n runTest
   where
-    assertClean :: TestM ()
-    assertClean = SQS.assertQueueEmpty
-
     runTest :: Assertion
     runTest = do
         setup <- s
-        void . flip runReaderT setup . runTestM $ h >> assertClean
```
Comment (ChrisPenner) [inline services/galley/test/integration/TestHelpers.hs]: Not really sure what this was supposed to gain us? It works to just wipe the queues after every test; but maybe this is slow?

File: services/galley/src/Galley/External/LegalHoldService.hs [hunk 5]
```diff
@@ -148,3 +167,15 @@ validateServiceKey pem = liftIO $ readPublicKey >>= \pk ->
 
     minRsaKeySize :: Int
     minRsaKeySize = 256 -- Bytes (= 2048 bits)
+
+-- Types
+
+-- | When receiving tokens from other services which are 'just passing through'
+-- it's error-prone useless extra work to parse and render them from JSON over and over again.
+-- We'll just wrap them with this to give some level of typesafety and a reasonable JSON
+-- instance
```
Comment (fisx) [inline services/galley/src/Galley/External/LegalHoldService.hs]: I think this is good for now, but I disagree with the "useless extra work to parse" part.  If we know it should be a legal hold device token, and those have internal structure, we should check that structure here before we pass it on internally.  This throws errors early on when they are still helpful.

Perhaps you can leave some sentence to this end in the haddocks?

File: services/galley/src/Galley/Intra/Client.hs (hunk 2 above)
Comment (fisx) [reply services/galley/src/Galley/Intra/Client.hs]: :+1: yes, it has to be zuid, or whatever name we have once https://github.com/wireapp/wire-server/pull/761 is ready.

File: services/galley/test/integration/API/Teams/LegalHold.hs [hunk 6]
```diff
@@ -495,12 +523,12 @@ getUserStatus uid tid = do
...
 
-approveLegalHoldDevice :: HasCallStack => UserId -> TeamId -> TestM ResponseLBS
-approveLegalHoldDevice uid tid = do
+approveLegalHoldDevice :: HasCallStack => UserId -> UserId -> TeamId -> TestM ResponseLBS
+approveLegalHoldDevice zusr uid tid = do
     g <- view tsGalley
     put $ g
-           . paths ["teams", toByteString' tid, "legalhold", "approve"]
-           . zUser uid . zConn "conn"
+           . paths ["teams", toByteString' tid, "legalhold", toByteString' uid, "approve"]
```
Comment (fisx) [inline services/galley/test/integration/API/Teams/LegalHold.hs]: hum, i deliberately removed the `uid` this from the path because it is redundant.  why did you add it again?  (not a big deal, just wondering.)

File: services/galley/test/integration/API/Teams/LegalHold.hs [hunk 7]
```diff
@@ -184,22 +186,48 @@ testCreateTwoLegalHoldDevices = do
...
+    (owner, tid) <- createTeam
+    member <- randomUser
+    addTeamMemberInternal tid $ newTeamMember member (rolePermissions RoleMember) Nothing
+
+    cannon <- view tsCannon
+
+    WS.bracketR cannon member $ \ws -> withDummyTestServiceForTeam owner tid $ do
+        -- not allowed to approve if team setting is disabled
+        -- TODO: remove the following 'ignore' once 'disabled' is the default
+        ignore $ approveLegalHoldDevice owner member tid !!! const 403 === statusCode
```
Comment (fisx) [inline services/galley/test/integration/API/Teams/LegalHold.hs]: can you test that this also happens if you:

> enable -> request device -> disable -> approve

File: services/galley/test/integration/API/Teams/LegalHold.hs [hunk 8]
```diff
@@ -184,22 +186,48 @@ testCreateTwoLegalHoldDevices = do
...
+        approveLegalHoldDevice owner member tid !!! const 403 === statusCode
+        approveLegalHoldDevice member member tid !!! const 200 === statusCode
 
-    when False $ void $ approveLegalHoldDevice undefined undefined
-      -- just something silly to keep the name to suppress the `-Wunused-top-binds` noise and
-      -- to remind everybody it's already there.
+        cassState <- view tsCass
+        liftIO $ do
+            clients' <- Cql.runClient cassState $ Data.lookupClients [member]
+            print clients'
```
Comment (fisx) [inline services/galley/test/integration/API/Teams/LegalHold.hs]: remove this line.

File: services/galley/test/integration/API/Teams/LegalHold.hs [hunk 9]
```diff
@@ -184,22 +186,48 @@ testCreateTwoLegalHoldDevices = do
...
+            assertBool "Expect clientId to be saved on the user"
+              $ Clients.contains member someClientId clients'
```
Comment (fisx) [inline services/galley/test/integration/API/Teams/LegalHold.hs]: it's a bit weird that this is a global variable, but i think i commited that crime at least once in this module before...  :)

File: libs/bilge/src/Bilge/Response.hs (hunk 1 above)
Comment (ChrisPenner) [reply libs/bilge/src/Bilge/Response.hs]: I tried that; the cookieJar itself is completely empty even though the `Set-Cookie` header is definitely there. It doesn't show up in `getCookie` either.

File: services/galley/src/Galley/Intra/Client.hs (hunk 2 above)
Comment (ChrisPenner) [reply services/galley/src/Galley/Intra/Client.hs]: Do you happen to know what the `CookieLabel` parameter on the request to `sso-login` is for? I must be misunderstanding its purpose then.

File: services/galley/test/integration/API/Teams/LegalHold.hs (hunk 9 above)
Comment (ChrisPenner) [reply services/galley/test/integration/API/Teams/LegalHold.hs]: The reason is because the clientId must match the fingerprint of the keys which we use (which are also a global variable) I added a comment on those saying they shouldn't be changed 😄

File: services/galley/test/integration/API/Teams/LegalHold.hs (hunk 6 above)
Comment (ChrisPenner) [reply services/galley/test/integration/API/Teams/LegalHold.hs]: I wasn't aware you'd removed it intentionally; I'm just following the spec on https://github.com/wearezeta/backend-issues/issues/826 so I just assumed this was a typo. 

Also; I really consider it bad API design to have conflicting routes in which one section may be static OR dynamic; e.g.:

- `teams/:tid/legalhold/:uid` 
- `teams/:tid/legalhold/approve`

Having both of these means that re-ordering the route matchers in the routing file WILL break the system, since `approve` will be matched to `:uid`. For this reason I'd prefer to keep it consistent 😄 

I do a check in the handler that the two uid's are the same, so it doesn't allow any strange behaviour 👍

File: services/galley/src/Galley/API/LegalHold.hs [hunk 10]
```diff
@@ -125,3 +128,33 @@ requestDevice (zusr ::: tid ::: uid ::: _) = do
...
+            Logger.warn lg $ Logger.msg @Text "No prekeys found"
+            throwM noLegalHoldDeviceAllocated
+        Just keys -> pure keys
+
+    clientId <- addLegalHoldClientToUser uid connId prekeys lastPrekey'
+    Data.updateClient True uid clientId
+
+    legalHoldAuthToken <- getLegalHoldAuthToken uid
+    LHService.confirmLegalHold clientId tid uid legalHoldAuthToken
+    LegalHoldData.setUserLegalHoldStatus uid UserLegalHoldEnabled
```
[the remark this answers is not in the document]
Comment (ChrisPenner) [reply services/galley/src/Galley/API/LegalHold.hs]: I thought about that, but decided against it; I can't think of any reason why I'd need to do that, and NOT doing it makes this handler more idempotent and less prone to getting into weird states.

If I delete it, but fail before responding, then the client will retry and the whole thing will fail.
If I leave them there, then the whole process *should* be mostly idempotent (though I may have missed a few edge cases, it's very tricky to test).

File: services/galley/test/integration/API/Teams/LegalHold.hs (hunk 7 above)
Comment (ChrisPenner) [reply services/galley/test/integration/API/Teams/LegalHold.hs]: Question; should disabling cause pending devices to be deleted? I'd assume it's easiest if we don't?

File: libs/bilge/src/Bilge/Response.hs (hunk 1 above)
Comment (fisx) [reply libs/bilge/src/Bilge/Response.hs]: no idea what's going on there either...

File: services/galley/src/Galley/Intra/Client.hs (hunk 2 above)
Comment (fisx) [reply services/galley/src/Galley/Intra/Client.hs]: after off-line discussion: may be documented in swagger.  it's used by clients to reference individual cookies they want to e.g. delete.

File: services/galley/src/Galley/API/LegalHold.hs (hunk 10 above)
[the remark this answers is not in the document]
Comment (fisx) [reply services/galley/src/Galley/API/LegalHold.hs]: cruft?

but you're right about the later failure when they get deleted and then the handler crashes.  that may be worth a comment in the code?

File: services/galley/test/integration/API/Teams/LegalHold.hs (hunk 6 above)
Comment (fisx) [reply services/galley/test/integration/API/Teams/LegalHold.hs]: very good point, about the hybrid dynamic-static path segment.

File: services/galley/test/integration/API/Teams/LegalHold.hs (hunk 7 above)
Comment (fisx) [reply services/galley/test/integration/API/Teams/LegalHold.hs]: not sure.  i would have tried to delete it, but can't think of a strong reason why we should.

**Comment by fisx** [comment]
re-approved!
